src/ApacheOrcDotNet.OptimizedReaderTest.App/TradeDataSource.cs

```diff
@@ -0,0 +1,127 @@
+            var timeColumn = _reader.GetColumn("time");
+            var priceColumn = _reader.GetColumn("price");
+            var sizeColumn = _reader.GetColumn("size");
+
+            _sourceColumnBuffer = reader.CreateStringColumnBuffer(sourceColumn);
+            _symbolColumnBuffer = reader.CreateStringColumnBuffer(symbolColumn);
+            _timeColumnBuffer = reader.CreateDecimalColumnBuffer(timeColumn);
+            _priceColumnBuffer = reader.CreateDecimalColumnBuffer(priceColumn);
+            _sizeColumnBuffer = reader.CreateIntegerColumnBuffer(sizeColumn);
+
+            var sourceFilterValues = FilterValues.CreateFromString(min: _souce, max: _souce);
+            var symbolFilterValues = FilterValues.CreateFromString(min: _symbol, max: _symbol);
+            var timeFilterValues = FilterValues.CreateFromTime(min: _startTime, max: _endTime);
+
+            var stripeIds = _reader.FilterStripes(sourceColumn, sourceFilterValues);
+            stripeIds = _reader.FilterStripes(stripeIds, symbolColumn, symbolFilterValues);
+            stripeIds = _reader.FilterStripes(stripeIds, timeColumn, timeFilterValues);
+
+            foreach (var stripeId in stripeIds)
+            {
+                var rowGroupIndexes = _reader.FilterRowGroups(stripeId, sourceColumn, sourceFilterValues);
+                rowGroupIndexes = _reader.FilterRowGroups(rowGroupIndexes, stripeId, symbolColumn, symbolFilterValues);
+                rowGroupIndexes = _reader.FilterRowGroups(rowGroupIndexes, stripeId, timeColumn, timeFilterValues);
+
+                _filters.Add(stripeId, rowGroupIndexes.ToList());
+            }
+        }
+
+        public int ApproxRowCount
+        {
+            get
+            {
+                var rowGroups = _filters.Values.Sum(v => v.Count);
+                return rowGroups * _reader.MaxValuesPerRowGroup;
+            }
+        }
+
+        public int ReadBatch(Span<decimal?> times, Span<decimal?> prices, Span<long?> sizes)
+        {
+            var numRows = 0;
+            var startTimeTotalSeconds = (decimal)_startTime.TotalSeconds;
+            var endTimeTotalSeconds = (decimal)_endTime.TotalSeconds;
+
+            foreach (var (stripeId, rowGroupIndexes) in _filters)
+            {
+                foreach (var rowEntryIndex in rowGroupIndexes)
+                {
+                    Task.WhenAll(
+                        _reader.LoadDataAsync(stripeId, rowEntryIndex, _sourceColumnBuffer),
+                        _reader.LoadDataAsync(stripeId, rowEntryIndex, _symbolColumnBuffer),
+                        _reader.LoadDataAsync(stripeId, rowEntryIndex, _timeColumnBuffer),
+                        _reader.LoadDataAsync(stripeId, rowEntryIndex, _sizeColumnBuffer),
+                        _reader.LoadDataAsync(stripeId, rowEntryIndex, _priceColumnBuffer)
+                    ).Wait();
+
+                    for (int idx = 0; idx < _reader.NumValuesLoaded; idx++)
```

Comment:
I'd like to spend just a little more time thinking about if we could pass our own buffers into `LoadDataAsync` rather than copying once more.

Reply:
The main issue with this is that we will not be able to use `Span` or `ref` for these as the method is `async`.